src/utils/size-observer.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ContentSizeObserverCallbackDef;
+
+/** @const {!WeakMap<!Window, !ResizeObserver>} */
+const observers = new WeakMap();
+
+/** @const {!WeakMap<!Element, !ContentSizeObserverCallbackDef>} */
+const targetObserverListMap = new WeakMap();
+
+/** @const {!WeakMap<!Element, !../layout-rect.LayoutSizeDef>} */
+const targetSizeListMap = new WeakMap();
+
+/**
+ * @param {!Element} element
+ * @param {!ContentSizeObserverCallbackDef} callback
+ */
+export function observeContentSize(element, callback) {
+  const win = element.ownerDocument.defaultView;
+  if (!win) {
+    return;
+  }
+  let callbacks = targetObserverListMap.get(element);
+  if (!callbacks) {
+    callbacks = [];
+    targetObserverListMap.set(element, callbacks);
+    getObserver(win).observe(element);
+  }
+  if (pushIfNotExist(callbacks, callback)) {
+    const size = targetSizeListMap.get(element);
+    if (size) {
+      setTimeout(() => callCallbackNoInline(callback, size));
+    }
+  }
+}
+
+/**
+ * @param {!Element} element
+ * @param {!ObserverCallbackDef} callback
+ */
+export function unobserveContentSize(element, callback) {
+  const callbacks = targetObserverListMap.get(element);
+  if (!callbacks) {
+    return;
+  }
+  removeItem(callbacks, callback);
+  if (callbacks.length == 0) {
+    targetObserverListMap.delete(element);
+    targetSizeListMap.delete(element);
+    const win = element.ownerDocument.defaultView;
+    if (win) {
+      getObserver(win).unobserve(element);
+    }
+  }
+}
+
+/**
+ * @param {!Element} element
+ * @return {!Promise<!../layout-rect.LayoutSizeDef>}
+ */
+export function measureContentSize(element) {
```

Comment:
I'm interested in the relative performance of this vs. InOb for the same usecase (size). Ideally ResizeObserver is faster since for the most part its getting _less_ information

Reply:
Hard to say. Most of my tests have shown me similar performance or InOb a bit faster. Importantly, however, `ResizeObserver` has additional rules when triggering can be delayed, while InOb always triggers "immediately". For instance `content-visibility:hidden` would block `ResizeObserver` callback until the subtree becomes "visible".